google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/GoogleComputeEngineServiceAdapter.java

```diff
@@ -70,223 +74,234 @@
+		this.api = api;
+		this.operationDone = operationDone;
+		this.instanceVisible = instanceVisible;
+		this.diskToSourceImage = diskToSourceImage;
+		this.resources = resources;
+		this.firewallTagNamingConvention = firewallTagNamingConvention;
+		this.imageProjects = Splitter.on(',').omitEmptyStrings().splitToList(imageProjects);
+	}
+
+
+	@Override public NodeAndInitialCredentials<Instance> createNodeWithGroupEncodedIntoName(String group, String name, Template template) {
+		GoogleComputeEngineTemplateOptions options = GoogleComputeEngineTemplateOptions.class.cast(template.getOptions());
+		checkNotNull(options.network(), "template options must specify a network");
+		checkNotNull(template.getHardware().getUri(), "hardware must have a URI");
+		checkNotNull(template.getImage().getUri(), "image URI is null");
+
+		List<AttachDisk> disks = Lists.newArrayList();
+		disks.add(AttachDisk.newBootDisk(template.getImage().getUri()));
+
+		disks.addAll(options.getDisks());
+		if (null != options.getAutoCreateDiskOptions()) {
+			AutoCreateDiskOptions diskOptions = options.getAutoCreateDiskOptions();
+			checkArgument(template.getLocation().getScope() == LocationScope.ZONE);
+			DiskApi diskApi = api.disksInZone(template.getLocation().getId());
+			Operation op = diskApi.create(diskOptions.getDiskName(name), new DiskCreationOptions.Builder().sizeGb(diskOptions.diskSizeGb).build());
+			// TODO: Is there a way to avoid the op, and just AttachDisk.create? where do we get the URI from? does it have to pre-exist?
+			AttachDisk disk = AttachDisk.create(diskOptions.diskType, diskOptions.diskMode, op.targetLink(), diskOptions.getDiskName(name), false, null, true, null, null);
+			disks.add(disk);
+		}
+
+		NewInstance newInstance = NewInstance.create(name, // name
+				template.getHardware().getUri(), // machineType
+				options.canIpForward(), //ip forward
+				options.network(), // network
+				disks, // disks
+				group // description
+				);
+		
+
+		// Add tags from template and for security groups
+		newInstance.tags().items().addAll(options.getTags());
+		FirewallTagNamingConvention naming = firewallTagNamingConvention.get(group);
+		for (int port : options.getInboundPorts()) {
+			newInstance.tags().items().add(naming.name(port));
+		}
+
+		// Add metadata from template and for ssh key and image id
+		newInstance.metadata().putAll(options.getUserMetadata());
+
+		LoginCredentials credentials = resolveNodeCredentials(template);
+		if (options.getPublicKey() != null) {
+			newInstance.metadata().put("sshKeys", format("%s:%s %s@localhost", credentials.getUser(), options.getPublicKey(), credentials.getUser()));
+		}
+
+		String zone = template.getLocation().getId();
+		InstanceApi instanceApi = api.instancesInZone(zone);
+		Operation create = instanceApi.create(newInstance);
+
+		// TODO: What does this AtomicReference do?
+		// TODO: What's the differenc between AttachDisk and Instance.AttachedDisk and can they be refactored together?
```

Comment:
Refactoring Instance.AttachedDisk and AttachDisk to be one object is a good idea. I'm not sure what this AtomicReference does, the comment is a little cryptic.